README.md
# ARM4S - Automatic Resource Management for Scala

 |Branch|Status|
 |:-----|:----:|
 |*master*|[![Build Status](https://travis-ci.com/tmoschou/arm4s.svg?branch=master)](https://travis-ci.com/tmoschou/arm4s)|
 |*develop*|[![Build Status](https://travis-ci.com/tmoschou/arm4s.svg?branch=develop)](https://travis-ci.com/tmoschou/arm4s)|

[![Scaladocs](https://www.javadoc.io/badge/io.tmos/arm4s_2.12.svg?label=Scaladoc)](https://www.javadoc.io/doc/io.tmos/arm4s_2.12)

This library provides a way of succinctly dealing with resources in an exception
safe manner. This library can provided identical exception handling and
execution semantics to Java's `try`-with-resource statement and a substitute for Scala
which does not have an equivalent construct natively.

More generally, managed resources are not limited to `java.lang.AutoClosable`s.
Any type with a **on-finally** or **on-exception** lifecycle, can be supported with
user-defined implicit adapters. For instance a `java.util.concurrent.ExecutorService`
may have a user-defined shutdown hook executed on-finally. See Comprehensive Example below.

Managed resources are treated as a singleton enumerator whose managed lifecycle is
scoped to an applied expression (or block). Unlike Java's constructs, ARM4S's
applied block can yield results, ensuring that resources are closed properly
before returning.

For example: Manage multiple resources and yield a result

```scala
import io.tmos.arm.Implicits._
import java.io._
import java.net.{Socket, InetAddress, ServerSocket}

val line = for {
  socket <- new Socket(InetAddress.getLoopbackAddress, port).manage
  out <- new PrintWriter(socket.getOutputStream, true).manage
  in <- new BufferedReader(new InputStreamReader(socket.getInputStream)).manage
} yield {
  val line = in.readLine()
  out.println(line.toUpperCase)
  return line
}
```

For more examples see, the Examples section below.

# Rationale

Manual management of resources have proven to be error prone, and when done
"correctly" - ugly.

Refer to
 * Joshua Bloch's Original
   [Proposal for ARM](http://mail.openjdk.java.net/pipermail/coin-dev/2009-February/000011.html).
 * Oracle's tech article on
   [Try-with-resources](http://www.oracle.com/technetwork/articles/java/trywithresources-401775.html).

For instance, if you are doing any of the following, then you should consider this library.

```scala
// don't do this
val r = new Resource(...)
try {
  doStuff1(r) // assume we throw an exception here
  doStuff2(r)
} finally {
  r.close() // what if we throw an exception here too?
}
```

Not good - we just masked (lost) the first 'important' exception with no
indication as to whether our main try block completed normally, or if it
did not, then whereabouts it did fail.

You may be tempted to wrap the close in another try/catch and log it
so that the first exception isn't ever dropped.

```scala
// don't do this either
val r = new Resource(...)
try {
  ...
} finally {
  try {
    r.close() // close quietly
  } catch {
    case e => log.warn(e)
  }
}
```

Still Bad - If `close()` throws an exception, the application has no idea one was thrown on close and
with no opportunity to fail fast and safely. Especially so, if the main try clause didn't
throw any exception at all, in which case _no_ exception is propagated.

Instead we should utilise `Throwable.addSuppressed` to propagate the 'first' important exception
 with any subsequent exceptions attached as 'suppressed'.

Now, that was for one resource. - What if you needed to close multiple resources
 in a finally block, each of which could independently throw an exception on close.
Could you get it right? If you do - well done.
But the next developer who reads is unlikely to understand it.

This is where this library comes in and does things *correctly* and *succinctly*,
ensuring that the first exception thrown is the one that
is propagated, and any subsequent exceptions thrown are added to the head exception as suppressed.

## Exception Behaviour
This library differs from other Scala ARM libraries in that it has been designed with consideration for different
exception scenarios and with the following goals regarding exception safe behaviour:

1. The `onFinally` (by default delegates to `close` for `java.lang.AutoClosables`s) and `onException` management
   hooks of a resource method must be called even if the body throws _any_ `Throwable` exception
   including fatal ones to ensure that no resources are leaked. Example of fatal exceptions include anything not
   matched by `scala.util.control.NonFatal` such as `InterruptedException`, `ControlThrowable` and
   `VirtualMachineError`. Though you should not try to handle such fatal errors, finally logic should still
   (attempted to) be executed regardless.

2. The `onFinally` and `onException` execution hooks are permitted to throw any `Throwable` too,
   possibly additional to exceptions thrown from the main block.

3. Any `Throwable` thrown by `onFinally` or `onException` should not mask any exception thrown firstly by the
   body, if any. Instead the secondary exception(s) thrown should be caught and recorded as a
   suppressed exception against the primary (currently throwing) exception.

4. Lastly we differ slightly to Java's implementation of `try`-with-resources in which we permit
   the corner case where exceptions thrown `onFinally` or `onException` may be the same instance
   as thrown by the applied expression. Where Java will throw a new `IllegalArgumentException` on
   attempts to self suppress, we will silently drop repeated instances as usually it is not the
   users fault that an underlying or decorated resource used a cached exception rather than
   generate a new exception/stacktrace etc.

## Including ARM4S in your project

In SBT:
```scala
libraryDependencies += "io.tmos" %% "arm4s" % "1.1.0"
```
In Maven:
```xml
<dependency>
    <groupId>io.tmos</groupId>
    <artifactId>arm4s_${scala.binary.version}</artifactId>
    <version>1.1.0</version>
</dependency>
```
## Using ARM4S

There are two ways you can construct a managed resource

Explicitly
```scala
import io.tmos.arm.ArmMethods._
for (r <- manage(resource)) {
  ...
}
```

Or using implicit decorator methods
```scala
import io.tmos.arm.Implicits._
for (r <- resource.manage) {
   ...
}
```

Any resource of type `T` for which an implicit `CanManage[T]` adapter is provided in scope can be managed.

By default the implicit `CloseOnFincally extends CanManage[AutoClosable]`
manager defined in the CanManage companion object is used, which calls `close`
on-finally, unless a higher priority implicit is in scope.
Alternatively `closeOnFinally` method may be used in place of `manage` method,
To explicitly use this adapter.

```scala
import io.tmos.arm.ArmMethods._
for (r <- closeOnFinally(resource)) {...}
```

or using the implicit method
```scala
import io.tmos.arm.Implicits._
for (r <- resource.closeOnFinally) {...}
```

Managed resources may be composed together/chained in a monadic manner that allows for optionally yielding
results or imperatively using `for`-comprehensions.

## Examples
Using
[For-Comprehensions](https://www.scala-lang.org/files/archive/spec/2.12/06-expressions.html#for-comprehensions-and-for-loops)
```scala
import io.tmos.arm.ArmMethods._
val jsonMap: Map[String, Any] = for {
  inputStream <- manage(new FileInputStream("data.json"))
} yield {
  JsonMethods.parse(inputStream).extract[Map[String, Any]]
}
```

We could also write in the following fluid style
```scala
import io.tmos.arm.Implicits._
val jsonMap: Map[String, Any] = new FileInputStream("data.json")
 .manage
 .map(JsonMethods.parse(_))
 .extract[Map[String, Any]]
```
Or if composing multiple resources this can be done easily too
```scala
import io.tmos.arm.ArmMethods._
val result = for {
  a <- manage(new A)
  b <- manage(a.getB)
  c <- manage(new C(b))
} yield {
  // ...
}
```

Resources will be safely managed in reverse declaration order even if a later enumerator
declaration threw an exception prior to the main body. For example if `new C(b)`
thew an exception, then `b` followed by `a`'s on-exception/on-finally lifecycle
hooks will be executed.

There may be cases where you need to override the default behaviour for AutoClosable and
close resources safely _only_ on-exception, such as when needing to construct
multiple resources atomically or delegating close to a different thread/scope.
This can be achieved using the predefined `CloseOnException` manager.

```scala
import io.tmos.arm.ArmMethods._

// this implicit now has higher priority then the default CloseOnFinally
implicit val canManage: CanManage[AutoCloseable] = CanManage.CloseOnException

def openAll(): (A, B, C) = for {
  a <- manage(new A)
  b <- manage(new B)
  c <- manage(new C)
} yield (a,b,c)
```

We also provided `closeOnException` method similarly to `closeOnFinally`, for
explicit usage of this adapter without needing to import it as a higher priority
implicit.

## Comprehensive Example

Here is a comprehensive (hypothetical) example of managing multiple resources implicitly,
including an `ExectorService` which we define `onFinally` logic for.
This sample code runs a server socket in a separate thread echoing back text it
receives in uppercase.

```scala
import io.tmos.arm.Implicits._
import scala.collection.JavaConverters._

implicit val manager: CanManage[ExecutorService] = new CanManage[ExecutorService] {
  override def onFinally(pool: ExecutorService): Unit = {
    pool.shutdown() // Disable new tasks from being submitted
    try {
      if (!pool.awaitTermination(10, TimeUnit.SECONDS)) { // wait for normal termination
        pool.shutdownNow() // force terminate
        if (!pool.awaitTermination(10, TimeUnit.SECONDS)) // wait for forced termination
          throw new RuntimeException("ExecutorService did not terminate")
      }
    } catch {
      case _: InterruptedException =>
        pool.shutdownNow() // (Re-)Cancel if current thread also interrupted
        Thread.currentThread().interrupt() // Preserve interrupt status
        // It is very important that we do not propagate InterruptedException
        // as it may be added as it may be suppressed if an earlier
        // exception trumps its. This is the same generally for any close method.
        // See https://docs.oracle.com/javase/8/docs/api/java/lang/AutoCloseable.html#close--
        // for details.
    }
  }
  override def onException(r: ExecutorService): Unit = {}
}

val serverSocketFuture = new CompletableFuture[ServerSocket]

val callable = new Callable[Unit] {
  override def call(): Unit = {
    // Note that ServerSocket.accept() blocks but, does not throw
    // InterruptedException. Instead, to terminate the event loop, we need
    // to close server socket asynchronously. We delegate closing of server
    // socket to the main thread under normal circumstances, but in the
    // event this thread has thrown an exception, we will close.
    for (ss <- new ServerSocket(0, 0, InetAddress.getLoopbackAddress).closeOnException) {
      serverSocketFuture.complete(ss)
      while (!Thread.interrupted()) { // main event loop
        try {
          for {
            connection <- ss.accept.closeOnFinally // block here
            out <- new PrintWriter(connection.getOutputStream, true).closeOnFinally
            in <- new BufferedReader(new InputStreamReader(connection.getInputStream)).closeOnFinally
            line <- in.lines().iterator().asScala
          } out.println(line.toUpperCase)
        } catch {
          case _: SocketException if ss.isClosed =>
            // at this point server socket has been closed via the main thread
            // we set interrupt status and terminate the event loop / thread
            Thread.currentThread().interrupt()
        }
      }
    }
  }
}

// manage an executor service using the user defined CanManage
val completedFuture = for (
  executorService <- Executors.newSingleThreadExecutor().manage
) yield {
  val future = executorService.submit(callable)
  for (ss <- serverSocketFuture.get.closeOnFinally) {
    val upperPhrase = for {
      s <- new Socket(InetAddress.getLoopbackAddress, ss.getLocalPort).closeOnFinally
      out <- new PrintWriter(s.getOutputStream, true).closeOnFinally
      in <- new BufferedReader(new InputStreamReader(s.getInputStream)).closeOnFinally
    } yield {
      out.println("hello")
      out.println("world")
      in.readLine() + ' ' + in.readLine()
    }
    assert(upperPhrase === "HELLO WORLD")
  }
  // at this point the callable event loop is terminating
  future
}
// at this point the callable event loop has been terminated

assert(!completedFuture.isCancelled)
assert(completedFuture.isDone)
completedFuture.get // should not block or throw any error
```


CHANGELOG.md
# Changelog

## 1.1.1 (Unreleased)
- Added cross build for Scala 2.13.1, 2.12.10

## [1.1.0](https://github.com/tmoschou/arm4s/releases/tag/v1.1.0) (2018-10-30)
- Added cross build for Scala 2.13-M5
- Support explicitly passing the CanManage object to `ImplicitManageable`'s methods [#5](https://github.com/tmoschou/arm4s/issues/5)
- `CanManage`'s `onFinally` and`onException` now defaults to no-op implementation [#6](https://github.com/tmoschou/arm4s/issues/6) 

## [1.0.0](https://github.com/tmoschou/arm4s/releases/tag/v1.0.0) (2018-09-27)
New API Improvements [#3](https://github.com/tmoschou/arm4s/pull/3): 
- CanManage/DefaultManagedResource now supports on-exception lifecycle hooks
- Implicit decorator methods to now explicitly convert a resource into a 
managed one, over implicit conversions.

## [0.2.0](https://github.com/tmoschou/arm4s/releases/tag/v0.2.0) (2017-01-30)
- Move implicits package-object to Implicits object. [#1](https://github.com/tmoschou/arm4s/issues/1) 

## [0.1.0](https://github.com/tmoschou/arm4s/releases/tag/v0.1.0) (2017-01-12)
- Initial release


src/test/scala/io/tmos/arm/ArmSuite.scala
package io.tmos.arm

import java.io._
import java.net.{Socket, InetAddress, ServerSocket}
import java.util.concurrent._

import org.scalatest.FunSuite

class ArmSuite extends FunSuite {

  private class SimpleAutoCloseableTest(val msg: String) extends Closeable {
    protected var closed = false
    def except = throw new RuntimeException(msg)
    def isClosed = closed
    override def close(): Unit = closed = true
  }

  implicit val canManageExectorService = new CanManage[ExecutorService] {
    override def withFinally(pool: ExecutorService): Unit = {
      pool.shutdown() // Disable new tasks from being submitted
      try {
        if (!pool.awaitTermination(10, TimeUnit.SECONDS)) { // wait for normal termination
          pool.shutdownNow() // force terminate
          if (!pool.awaitTermination(10, TimeUnit.SECONDS)) // wait for forced termination
            throw new RuntimeException("ExecutorService did not terminate")
        }
      } catch {
        case e: InterruptedException =>
          pool.shutdownNow()  // (Re-)Cancel if current thread also interrupted
          Thread.currentThread().interrupt()  // Preserve interrupt status
      }
    }
  }

  test("Resources should be closed when explicitly managed") {
    val resource: SimpleAutoCloseableTest = new SimpleAutoCloseableTest("1")
    for (r <- manage(resource)) {
      assert(!r.isClosed)
    }
    assert(resource.isClosed)
  }

  test("Resources should be closed when implicitly managed") {
    import io.tmos.arm.implicits._
    val resource: SimpleAutoCloseableTest = new SimpleAutoCloseableTest("1")
    for (r <- resource) {
      assert(!r.isClosed)
    }
    assert(resource.isClosed)
  }

  test("Resources should be closed in reverse order when foreach-ing over them") {
    var resource1: SimpleAutoCloseableTest = null
    var resource2: SimpleAutoCloseableTest = null
    var resource3: SimpleAutoCloseableTest = null

    resource1 = new SimpleAutoCloseableTest("1") {
      override def close(): Unit = {
        super.close()
        assert(resource2.isClosed)
        assert(resource3.isClosed)
      }
    }

    resource2 = new SimpleAutoCloseableTest("2") {
      override def close(): Unit = {
        super.close()
        assert(!resource1.isClosed)
        assert(resource3.isClosed)
      }
    }

    resource3 = new SimpleAutoCloseableTest("3") {
      override def close(): Unit = {
        super.close()
        assert(!resource1.isClosed)
        assert(!resource2.isClosed)
      }
    }

    for {
      r1 <- manage(resource1)
      r2 <- manage(resource2)
      r3 <- manage(resource3)
    } {
      assert(!r1.isClosed)
      assert(!r2.isClosed)
      assert(!r3.isClosed)
    }

    assert(resource1.isClosed)
    assert(resource2.isClosed)
    assert(resource3.isClosed)
  }

  test("Resources should be closed in reverse order when yield-ing over them") {
    var resource1: SimpleAutoCloseableTest = null
    var resource2: SimpleAutoCloseableTest = null
    var resource3: SimpleAutoCloseableTest = null

    resource1 = new SimpleAutoCloseableTest("1") {
      override def close(): Unit = {
        super.close()
        assert(resource2.isClosed)
        assert(resource3.isClosed)
      }
    }

    resource2 = new SimpleAutoCloseableTest("2") {
      override def close(): Unit = {
        super.close()
        assert(!resource1.isClosed)
        assert(resource3.isClosed)
      }
    }

    resource3 = new SimpleAutoCloseableTest("3") {
      override def close(): Unit = {
        super.close()
        assert(!resource1.isClosed)
        assert(!resource2.isClosed)
      }
    }

    val value = for {
      r1 <- manage(resource1)
      r2 <- manage(resource2)
      r3 <- manage(resource3)
    } yield r1.msg + r2.msg + r3.msg

    assert(resource1.isClosed)
    assert(resource2.isClosed)
    assert(resource3.isClosed)
    assert(value === "123")
  }

  test("Resource should be closed when exception is thrown") {
    val resource = new SimpleAutoCloseableTest("msg")
    try {
      for { r <- manage(resource) }
        r.except
    } catch {
      case e: RuntimeException =>
        assert(e.getMessage === "msg")
    }
    assert(resource.isClosed)
  }

  test("Correct exception should be propagated when close on resource also excepts whilst currently excepting") {
    val resource = new SimpleAutoCloseableTest("msg1") {
      override def close(): Unit = {
        super.close()
        throw new RuntimeException("msg2")
      }
    }
    try {
      for { r <- manage(resource) }
        r.except
    } catch {
      case e: RuntimeException =>
        assert(e.getMessage === "msg1")
        val suppressed = e.getSuppressed
        assert(suppressed.length === 1)
        assert(suppressed(0).getMessage === "msg2")
    }
    assert(resource.isClosed)
  }

  test("Composing resources and lookup of user defined CanManage should work") {

    import io.tmos.arm.implicits._
    import scala.collection.JavaConverters._

    val port = new CompletableFuture[Int]

    val callable = new Callable[Unit] {
      override def call(): Unit = {
        for (ss <- new ServerSocket(0, 0, InetAddress.getLoopbackAddress)) {
          port.complete(ss.getLocalPort)
          for {
            connection <- ss.accept
            out <- new PrintWriter(connection.getOutputStream, true)
            in <- new BufferedReader(new InputStreamReader(connection.getInputStream))
            line <- in.lines().iterator().asScala // note not a resource, but now a traversable
          } out.println(line.toUpperCase)
        }
      }
    }

    // manage an executor service using the user defined canManageExectorService.
    // This will shutdown and wait termination
    val completedFuture = for (executorService <- Executors.newSingleThreadExecutor()) yield {
      val future = executorService.submit(callable)

      val upperPhrase = for {
        s <- new Socket(InetAddress.getLoopbackAddress, port.get)
        out <- new PrintWriter(s.getOutputStream, true)
        in <- new BufferedReader(new InputStreamReader(s.getInputStream))
      } yield {
        out.println("hello")
        out.println("world")
        in.readLine() + ' ' + in.readLine()
      }
      assert(upperPhrase === "HELLO WORLD")
      future
    }

    assert(completedFuture.isCancelled || completedFuture.isDone)
    completedFuture.get() // should not block or throw any error

  }

}



src/test/scala-2.12/io/tmos/arm/Compat.scala
package io.tmos.arm

object Compat {
  val CollectionConverters = scala.collection.JavaConverters
}


src/test/scala-2.13/io/tmos/arm/Compat.scala
package io.tmos.arm

object Compat {
  val CollectionConverters = scala.jdk.CollectionConverters
}


src/main/scala/io/tmos/arm/CanManage.scala
package io.tmos.arm

/**
  * For encapsulating the management logic of a resource.
  *
  * Default logic for any `java.lang.AutoClosable` is provided by the companion object,
  * which may be imported into current scope as implicits.
  *
  * Other types may be provided in scope by the user. For example
  * {{{
  *   import java.util.concurrent._
  *   import io.tmos.arm.Implicits._
  *
  *   implicit val manager: CanManage[ExecutorService] = new CanManage[ExecutorService] {
  *     override def onFinally(pool: ExecutorService): Unit = {
  *       pool.shutdown() // Disable new tasks from being submitted
  *       try {
  *         if (!pool.awaitTermination(10, TimeUnit.SECONDS)) { // wait for normal termination
  *           pool.shutdownNow() // force terminate
  *           if (!pool.awaitTermination(10, TimeUnit.SECONDS)) // wait for forced termination
  *             throw new RuntimeException("ExecutorService did not terminate")
  *         }
  *       } catch {
  *         case _: InterruptedException =>
  *           pool.shutdownNow() // (Re-)Cancel if current thread also interrupted
  *           Thread.currentThread().interrupt() // Preserve interrupt status
  *       }
  *     }
  *     override def onException(r: ExecutorService): Unit = {}
  *   }
  *
  *   for (manage(executorService) <- Executors.newSingleThreadExecutor.manage) { ... }
  * }}}
  *
  * @tparam R the type of the resource to manage
  */
trait CanManage[-R] {

  /**
    * Execution hook called after the managed block.
    *
    * This execution hook is called regardless if an exception is thrown.
    *
    * Usually resources are released or closed in the lifecycle.
    *
    * Implementors are free to permit exceptions thrown from this method, however
    * it is strongly advised to not have the
    * method throw `java.lang.InterruptedException`. This exception interacts with a thread's
    * interrupted status, and runtime misbehavior is likely to occur if an `java.lang.InterruptedException`
    * is suppressed. More generally, if it would cause problems for an exception to be suppressed,
    * the AutoCloseable.close method should not throw it."
    *
    * @param r the resource being managed
    */
  def onFinally(r: R): Unit = {}

  /**
    * Execution hook called when an exception is thrown from the managed
    * block. This is executed prior to [onFinally].
    *
    * Implementors are free to permit exceptions thrown from this method, however
    * note that any new exceptions thrown will be added as
    * a suppressed exception of the currently throwing exception.
    * Thus it is strongly advised that implementors do not throw any exceptions
    * if it would cause problems for an exception  to be suppressed.
    *
    * @param r the resource being managed
    */
  def onException(r: R): Unit = {}
}


/**
 * Companion object to the CanManage type trait.
 *
 * Contains common implementations of CanManage for AutoClosable Resources
 */
object CanManage {

  /**
    * Always call close on a AutoClosable after applied block,
    * regardless if block throws an exception or not. This is identical to
    * Java's try-with-resources.
    */
  implicit object CloseOnFinally extends CanManage[AutoCloseable] {
    override def onFinally(r: AutoCloseable): Unit = if (r != null) r.close()
  }

  /**
    * Call close on a resource only if a exception is thrown in the applied block.
    * This is useful for instance if closing a resource needs to be delegated
    * elsewhere under normal circumstances, but abnormal circumstances should be
    * handled in the current scope. Such examples may include managing a resource
    * across threads.
    */
  object CloseOnException extends CanManage[AutoCloseable] {
    override def onException(r: AutoCloseable): Unit = if (r != null) r.close()
  }

}


src/main/scala/io/tmos/arm/Implicits.scala
package io.tmos.arm

/**
  * Implicit Methods for management of a resource.
  *
  * For explicit management see `io.tmos.arm.ArmMethods`.
  */
object Implicits {

  /**
    * Provide implicit methods to manage a generic resource.
    *
    * Also requires an implicit `CanManage[S]` in scope. A default manager for
    * `AutoClosable`s is provided (See `CanManage.CloseOnFinally`) and the
    * default behaviour is equivalently the same as `closeOnFinally`
    *
    * @param r the resource to managed passed by-name
    * @tparam R the type of the resource passed to the applied expression
    */
  implicit class ImplicitManageable[R](r: =>  R) {

    /**
      * Converts this resource into a generic managed resource.
      *
      * @tparam S the type of the resource passed to the manager
      * @return the managed resource
      */
    def manage[S >: R](implicit canManage: CanManage[S]): ManagedResource[R] = ArmMethods.manage(r)
  }

  /**
    * Provide implicit methods to manage an AutoClosable resource.
    *
    * @param r the resource to managed passed by-name
    * @tparam R the type of the resource passed to the applied expression
    */
  implicit class ImplicitAutoClosable[R <: AutoCloseable](r: => R) {

    /**
      * Convert this AutoClosable into a managed resource. Will call
      * `AutoClosable.close` during the on-finally execution lifecycle.
      * The on-exception lifecycle hook is a no-op.
      *
      * @return the managed resource
      */
    def closeOnFinally: ManagedResource[R] = ArmMethods.closeOnFinally(r)

    /**
      * Convert this AutoClosable into a managed resource. Will call
      * `AutoClosable.close` during the on-exception execution lifecycle.
      * The on-finally lifecycle hook is a no-op.
      *
      * @return the managed resource
      */
    def closeOnException: ManagedResource[R] = ArmMethods.closeOnException(r)
  }

}


src/main/scala/io/tmos/arm/package.scala
package io.tmos

/**
 * ARM4S - Automatic Resource Management for Scala
 *
 * This library provides a way of succinctly dealing with resources in an exception safe manner. The behaviour
 * is identical to Java's exception handling in `try`-with-resources statement and a substitute for scala which does not
 * have an equivalent construct natively.
 *
 * Unlike Java's try-with-resource constructs, managed resources are not limited to `java.lang.AutoClosable`s.
 * Further types such as `java.util.concurrent.ExecutorService` can be supported with implicit converters in user code.
 * See examples below. Also unlike try-with-resource constructs, ARM4S's managed block (the body) can
 * actually yield results, ensuring that resources are closed properly before returning.
 *
 * For example: Parse multiple resources implicitly and yield a result
 * {{{
 *     import io.tmos.arm.Implicits._
 *     import java.io._
 *     import java.net.{Socket, InetAddress, ServerSocket}
 *
 *     val line = for {
 *       socket <- new Socket(InetAddress.getLoopbackAddress, port)
 *       out <- new PrintWriter(socket.getOutputStream, true)
 *       in <- new BufferedReader(new InputStreamReader(socket.getInputStream))
 *     } yield {
 *       val line = in.readLine()
 *       out.println(line.toUpperCase)
 *       return line
 *     }
 * }}}
 * For more examples see, the Examples section below.
 *
 * == Exception Behaviour ==
 * This library differs from other Scala ARM libraries in that it has been designed with consideration for different
 * exception scenarios and with the following goals regarding exception safe behaviour:
 *
 * 1. The [[io.tmos.arm.CanManage.withFinally `withFinally`]] of a managed resource
 *    (delegates to `close` of `AutoCloseable`s) must be called even if the body throws ''any'' `Throwable` exception
 *    including fatal ones to ensure that no resources are leaked. Example of fatal exceptions include anything not
 *    matched by [[scala.util.control.NonFatal$]] such as `InterruptedException`, `ControlThrowable` and
 *    `VirtualMachineError`. Though you should not try to handle such unchecked errors, finally logic should still
 *    (attempted to) be executed regardless.
 *
 * 2. The `withFinally` method may also throw any `Throwable` too. Unfortunately this is a
 *    possibility as permitted by `AutoCloseable`, but gives flexibility in implementing cleanup logic.
 *
 * 3. Importantly, any `Throwable` thrown by `withFinally` should not mask (suppress) any exception thrown firstly by the
 *    body, if any. Instead the secondary exception thrown in the finally clause should be caught and recorded as a
 *    suppressed exception against the primary (currently throwing) exception. This is what Java's try with resource
 *    construct does; for more details, see Oracle's tech article on
 *    [[http://www.oracle.com/technetwork/articles/java/trywithresources-401775.html Try-with-resources]].
 *
 * == Using ARM4S ==
 *
 * There are two ways you can import arm4s.
 *
 * For explicit resource management
 * {{{
 *     import io.tmos.arm._
 *     for (r <- manage(resource))
 *       ...
 * }}}
 * Or implicitly
 * {{{
 *     import io.tmos.arm.Implicits._
 *     for (r <- resource)
 *        ...
 * }}}
 * Any resource of type `T` for which an implicit `CanManage[T]` object is provided in scope can be managed.
 *
 * By default the following `CanManage[T]` that are automatically provided in scope on import
 *   * `type T = java.lang.AutoClosable`
 *   * `type T = { def close() }` - via scala reflection
 *
 * Managed resources may be composed together/chained in a monadic manner that allows for optionally yielding
 * results or imperatively using `for`-comprehensions.
 *
 * The managed resources are automatically closed, and in reverse declaration order.
 *
 * The resources are closed in a `finally` clause regardless of any exception thrown in the body of the
 * `for`-comprehension, or any prior `withFinally` called on other resources.
 *
 * == Examples ==
 * Using [[https://www.scala-lang.org/files/archive/spec/2.12/06-expressions.html#for-comprehensions-and-for-loops
 * For-Comprehensions]]
 * {{{
 *     import io.tmos.arm._
 *     val lines: Seq[String] = for (inputStream <- managed(new FileInputStream("data.json")) yield {
 *       Json.parse(inputStream).as[Seq[String]]
 *     }
 * }}}
 * which translates the the following monadic style
 * {{{
 *     import io.tmos.arm._
 *     val lines = managed(new FileInputStream("data.json")) map { inputStream =>
 *        Json.parse(inputStream).as[Seq[String]]
 *     }
 * }}}
 * or if composing multiple resources this can be done easily too
 * {{{
 *     import io.tmos.arm._
 *     val result = for {
 *       a <- managed(new A)
 *       b <- managed(new B(a))
 *       c <- managed(new C)
 *     } yield {
 *       ...
 *     }
 * }}}

 * == Comprehensive Example ==
 *
 * Here is a comprehensive example of managing multiple resources implicitly, including an `ExectorService` which we
 * define `withFinally` logic for. This sample code runs a server socket in a separate thread echoing back text it
 * receives in uppercase.
 * {{{
 *     import io.tmos.arm.Implicits._
 *     import scala.collection.JavaConverters._
 *
 *     implicit val canManageExectorService = new CanManage[ExecutorService] {
 *       override def withFinally(pool: ExecutorService): Unit = {
 *         pool.shutdown() // Disable new tasks from being submitted
 *         try {
 *           if (!pool.awaitTermination(10, TimeUnit.SECONDS)) { // wait for normal termination
 *             pool.shutdownNow() // force terminate
 *             if (!pool.awaitTermination(10, TimeUnit.SECONDS)) // wait for forced termination
 *               throw new RuntimeException("ExecutorService did not terminate")
 *           }
 *         } catch {
 *           case e: InterruptedException =>
 *             pool.shutdownNow()  // (Re-)Cancel if current thread also interrupted
 *             Thread.currentThread().interrupt()  // Preserve interrupt status
 *         }
 *       }
 *     }
 *
 *     val port = new CompletableFuture[Int]
 *
 *     val callable = new Callable[Unit] {
 *       override def call(): Unit = {
 *         for (ss <- new ServerSocket(0, 0, InetAddress.getLoopbackAddress)) {
 *           port.complete(ss.getLocalPort)
 *           for {
 *             connection <- ss.accept
 *             out <- new PrintWriter(connection.getOutputStream, true)
 *             in <- new BufferedReader(new InputStreamReader(connection.getInputStream))
 *             line <- in.lines().iterator().asScala // note not a resource, but now a traversable
 *           } out.println(line.toUpperCase)
 *         }
 *       }
 *     }
 *
 *     // manage an executor service using the user defined canManageExectorService.
 *     // This will shutdown and wait termination
 *     val completedFuture = for (executorService <- Executors.newSingleThreadExecutor()) yield {
 *       val future = executorService.submit(callable)
 *       val upperPhrase = for {
 *         s <- new Socket(InetAddress.getLoopbackAddress, port.get)
 *         out <- new PrintWriter(s.getOutputStream, true)
 *         in <- new BufferedReader(new InputStreamReader(s.getInputStream))
 *       } yield {
 *         out.println("hello")
 *         out.println("world")
 *         in.readLine() + ' ' + in.readLine()
 *       }
 *       assert(upperPhrase === "HELLO WORLD")
 *       future
 *     }
 *
 *     assert(completedFuture.isCancelled || completedFuture.isDone)
 *     completedFuture.get() // should not block
 * }}}
 */
package object arm {

  /**
   * Explicitly manages a resource. For implicit management see `io.tmos.arm.Implicits`
   *
   * Note that the resource is passed by-name and as it not evaluated until the construct is applied
   * over the body. This allows for example resources to be constructed/open lazily.
   *
   * @param r the resource to managed passed by-name
   * @tparam R the type of the Resource
   *
   * @return a managed resource
   */
  def manage[R: CanManage](r: => R): ManagedResource[R] = new DefaultManagedResource(r)


}
